pkg/odo/cli/list/services/services.go

```diff
@@ -0,0 +1,183 @@
+	"os"
+	"strings"
+)
+
+const RecommendedCommandName = "services"
+
+var (
+	listExample = ktemplates.Examples(`
+	# List the bindable Operator backed services from current namespace
+    %[1]s
+
+	# List all the bindable Operator backed services from all the namespaces
+	%[1]s --all-namespaces
+	%[1]s -A
+
+	# List the bindable Operator backed services in JSON format
+	%[1]s -o json
+	%[1]s --all-namespaces -o json
+	%[1]s -A -o json`)
+
+	listLongDesc = ktemplates.LongDesc(`
+	List the bindable Operator backed services that could bound to the odo component 
+`)
+)
+
+type ServiceListOptions struct {
+	// clientset
+	clientset *clientset.Clientset
+	// context
+	*genericclioptions.Context
+	// working directory
+	contextDir string
+	// flags
+	namespaceFlag     string
+	allNamespacesFlag bool
+}
+
+var _ genericclioptions.Runnable = (*ServiceListOptions)(nil)
+var _ genericclioptions.JsonOutputter = (*ServiceListOptions)(nil)
+
+func (o *ServiceListOptions) SetClientset(clientset *clientset.Clientset) {
+	o.clientset = clientset
+}
+
+func (o *ServiceListOptions) Complete(cmdline cmdline.Cmdline, _ []string) error {
+	var err error
+	o.contextDir, err = os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	o.Context, err = genericclioptions.New(genericclioptions.NewCreateParameters(cmdline))
+	if err != nil {
+		return err
+	}
+
+	if o.namespaceFlag == "" && !o.allNamespacesFlag {
+		o.namespaceFlag = o.GetProject()
+	}
+
+	return err
+}
+
+func (o *ServiceListOptions) Validate() error {
+	if o.allNamespacesFlag && o.namespaceFlag != "" {
+		return errors.New("cannot use --all-namespaces and --namespace flags together")
+	}
+	return nil
+}
+
+func (o *ServiceListOptions) Run(_ context.Context) error {
+	var s *log.Status
+	if o.allNamespacesFlag {
+		s = log.Spinner("Listing bindable services from all namespaces")
+		defer s.End(false)
+	} else {
+		s = log.Spinner(fmt.Sprintf("Listing bindable services from namespace %q", o.namespaceFlag))
+		defer s.End(false)
+	}
+	services, err := o.run()
+	if err != nil {
+		return err
+	}
+	s.End(true)
+	HumanReadable(services)
+	return nil
+}
+
+func (o *ServiceListOptions) run() (api.ResourcesList, error) {
+	services := []unstructured.Unstructured{}
+
+	if o.allNamespacesFlag {
+		projects, err := o.clientset.ProjectClient.List()
+		if err != nil {
+			return api.ResourcesList{}, err
+		}
+		for _, project := range projects.Items {
+			svcs, err := o.clientset.BindingClient.GetServiceInstances(project.Name)
+			if err != nil {
+				return api.ResourcesList{}, err
+			}
+			for k := range svcs {
+				services = append(services, svcs[k])
+			}
+		}
+	} else {
+		svcs, err := o.clientset.BindingClient.GetServiceInstances(o.namespaceFlag)
+		if err != nil {
+			return api.ResourcesList{}, err
+		}
+		for k := range svcs {
+			services = append(services, svcs[k])
+		}
+	}
+
+	var servicesList []api.BindableService
+	for _, svc := range services {
+		servicesList = append(servicesList, api.BindableService{
```

Comment:
Why return this compacted name? Is it imposed?

It would be more open for user to have separated info (namespace, name, kind, group), or at least (namespace, name, kind+group).

Reply:
Yeah, that's from the acceptance criteria.

Reply:
IIUC, this compacted name is the value that should be passed as service to `odo add binding --service myservice/Redis.redis.redis.opstreelab.in --name myRedisService`. 


If it is the case, it would be interesting to have specific a function for building this string near the definition of `parseServiceName` in `pkg/binding/backend`

Also, it would be more understandable if the field is named `service` instead of `name`, to indicate it is the value of the `--service` flag.

To sum up, I think having those fields would be interesting:
- service: the value to pass to `add binding`
- name: the name of the instance
- namespace: 
- kind: the kind of the resource
- group/version: the group/version of the resource

So the plugin can display any of the name / namespace / kind / groupVersion to the user for them to choose one, and use service for `add binding`

Reply:
@kadel PTAL above comment and suggest.

Reply:
> Also, it would be more understandable if the field is named service instead of name, to indicate it is the value of the --service flag.

This is good point. `service` makes more sense here.


> To sum up, I think having those fields would be interesting:
> 
> service: the value to pass to add binding
> name: the name of the instance
> namespace:
> kind: the kind of the resource
> group/version: the group/version of the resource


We can definitely do this for JSON  output.
```json
[ {
    "name": "hippo",
    "service": "hippo/PerconaXtraDBCluster.pxc.percona.com",
    "kind": "PerconaXtraDBCluster",
    "group": "pxc.percona.com",
    "namespace": "asdf"
},{

    "name": "cluster1",
    "service": "cluster1.PostgresCluster.postgres-operator.crunchydata.com",
    "kind": "PostgresCluster",
    "group": "postgres-operator.crunchydata.com",
    "namespace": "asdf"
}]
```

but for normal output, I'm afraid that it will be too many columns.

```
 NAME      SERVICE                                                     KIND                  GROUP                               NAMESPACE
 hippo     hippo/PerconaXtraDBCluster.pxc.percona.com                  PerconaXtraDBCluster  postgres-operator.crunchydata.com   asdf
 cluster1  cluster1/PostgresCluster.postgres-operator.crunchydata.com  PostgresCluster       postgres-operator.crunchydata.com   asdf
```

For normal output I would prefer keeping as it is, only renaming name to service:

```
 SERVICE                                                     NAMESPACE
 hippo/PerconaXtraDBCluster.pxc.percona.com                  asdf
 cluster1/PostgresCluster.postgres-operator.crunchydata.com  asdf
```

Or we could consider both NAME, and SERVICE, 
```
 NAME      SERVICE                                                    NAMESPACE
 hippo     hippo/PerconaXtraDBCluster.pxc.percona.com                 asdf
 cluster1  cluster1/PostgresCluster.postgres-operator.crunchydata.com asdf
```

Reply:
For me, it is also interesting to have this `service` value in the JSON output, so the plugin does not have to compute this value when it wants to call `add binding` (I consider this as an "opaque token")

Reply:
I haven't added the `service` field to the commit I pushed. IMO the CLI layer and plugin should handle it by themselves instead. Here's the output:
```sh
$ odo list services -A
 ✓  Listing bindable services from all namespaces [57ms]

 NAME                                                   NAMESPACE  
 cluster-sample/Cluster.postgresql.k8s.enterprisedb.io  myproject  
 hello-world/RabbitmqCluster.rabbitmq.com               newproject 

$ odo list services -A -o json
{
	"bindableServices": [
		{
			"name": "cluster-sample",
			"namespace": "myproject",
			"kind": "Cluster",
			"group": "postgresql.k8s.enterprisedb.io"
		},
		{
			"name": "hello-world",
			"namespace": "newproject",
			"kind": "RabbitmqCluster",
			"group": "rabbitmq.com"
		}
	]
}
```

Reply:
I don't agree, at it is error prone for the plugin to build this string by themselves (and not really their responsability).

Reply:
> For me, it is also interesting to have this `service` value in the JSON output, so the plugin does not have to compute this value when it wants to call `add binding` (I consider this as an "opaque token")

For me, this is precisely the reason we should not do it. I know that `run` function is at CLI layer, not at business layer, but producing JSON outputs that require consumers (here plugins) to not have to do anything except use a specific field is a bad idea. It would set a precedent for further requests.

If we would still like to do it, fine by me.

Reply:
> I don't agree, at it is error prone for the plugin to build this string by themselves (and not really their responsability).

I disagree. It's only string concatenation that needs to be done here. How could that be error-prone?

About responsibility, I think that JSON outputs should help the user have everything they need, which is surely the case here. By putting the service name in JSON output, we are spoon-feeding the users to not have to do anything at all. And that, in my opinion, is a bad precedent which could easily lead to more such requests from the users.

Reply:
I would like to understand your argument. It is not clear for me.

Mine is that a data structure generally contains a unique (and sometimes opaque) ID (in databases, in JSON/YAML structures like devfile, etc). Other data is here for display usage or for other usage, but not for refeencing the element. The unique ID is the only field used for this identification.

Here, the service would play the role of the unique ID.

Reply:
>By putting the service name in JSON output, we are spoon-feeding the users to not have to do anything at all.

We actually should spoon-feed them as much as we can (where it makes sense).


For me, the biggest argument is that if we show the full service name (service identifier) as it is required by `odo link` command, it will allow us to change the way how it is generated without breaking any integrations build on top of that.

Another argument is that the way how the service identifier is constructed is odo specific thing, as such it needs to be handled by odo, if we let plugins construct it on their side, then we are leaking odo internal logic.

If the service identifier string would be some kind of standard that is well defined and used somewhere else. Then it would make sense to just provide all required info and let others construct it on their own. But that is not the case here.

Reply:
TL;DR - 501e5a0

Responses below.

> if we show the full service name (service identifier) as it is required by `odo link` command, it will allow us to change the way how it is generated without breaking any integrations build on top of that.
> 
> Another argument is that the way how the service identifier is constructed is odo specific thing, as such it needs to be handled by odo, if we let plugins construct it on their side, then we are leaking odo internal logic.

I was thinking exactly this after my last response about concatenation and having every piece of info already there in the JSON output. And it inclined me to put the service field in there, as @feloy suggested.

> We actually should spoon-feed them as much as we can (where it makes sense).

I look at users as the odo CLI users and the plugin devs. For the first, I agree to spoon-feed and beyond! But for plugin devs, I think we should draw a line. However, in this case, it makes sense to me to put the service field.

> Mine is that a data structure generally contains a unique (and sometimes opaque) ID (in databases, in JSON/YAML structures like devfile, etc). Other data is here for display usage or for other usage, but not for refeencing the element. The unique ID is the only field used for this identification.
> 
> Here, the service would play the role of the unique ID.

I think you are saying this based on your experience and/or a paradigm you have come across. I wasn't looking at the idea of putting the service field in this way. Thanks for this info.